strawberryfields/io/xir_io.py

```diff
@@ -0,0 +1,294 @@
+code.
+"""
+# pylint: disable=protected-access,too-many-nested-blocks
+from decimal import Decimal
+from typing import Iterable
+
+import numpy as np
+
+import xir
+from xir.program import Declaration
+
+import strawberryfields.program as sfp
+from strawberryfields.tdm.tdmprogram import TDMProgram
+import strawberryfields.parameters as sfpar
+from strawberryfields import ops
+
+
+# pylint: disable=too-many-branches
+def from_xir(xir_prog):
+    """Convert an XIR Program to a Strawberry Fields program.
+
+    Args:
+        xir_prog (xir.Program): the input XIR program object
+
+    Returns:
+        Program: corresponding Strawberry Fields program
+    """
+    # only script-level statements are part of `xir_prog.statements`, which can only have integer
+    # wires, leading to `xir_prog.wires` only containing integer wire labels
+    num_of_modes = int(max(xir_prog.wires or [-1])) + 1
+    name = xir_prog.options.get("name", "xir")
+    if num_of_modes == 0:
+        raise ValueError(
+            "The XIR program is empty and cannot be transformed into a Strawberry Fields program"
+        )
+    prog = sfp.Program(num_of_modes, name=name)
+
+    # append the quantum operations
+    with prog.context as q:
+        for op in xir_prog.statements:
+            # check if operation name is in the list of
+            # defined StrawberryFields operations.
+            # This is used by checking against the ops.py __all__
+            # module attribute, which contains the names
+            # of all defined quantum operations
+            if op.name in ops.__all__:
+                # get the quantum operation from the sf.ops module
+                gate = getattr(ops, op.name)
+            else:
+                raise NameError(f"Quantum operation {op.name!r} not defined!")
+
+            # create the list of regrefs
+            regrefs = [q[i] for i in op.wires]
+
+            if op.params:
+                # convert symbolic expressions to symbolic expressions containing the corresponding
+                # MeasuredParameter and FreeParameter instances.
+                if isinstance(op.params, dict):
+                    vals = sfpar.par_convert(op.params.values(), prog)
+                    params = dict(zip(op.params.keys(), vals))
+                    gate(**params) | regrefs  # pylint:disable=expression-not-assigned
+                else:
+                    params = []
+                    for p in op.params:
+                        if isinstance(p, Decimal):
+                            params.append(float(p))
+                        elif isinstance(p, Iterable):
+                            params.append(np.array(_listr(p)))
+                        else:
+                            params.append(p)
+                    params = sfpar.par_convert(params, prog)
+                    gate(*params) | regrefs  # pylint:disable=expression-not-assigned
+            else:
+                if callable(gate):
```

Comment:
TIL about `callable` too! Is there a test where this `if` statement evaluates to `True`?

Reply:
Actually, it always evaluates to True, since the gate should always be callable.